### PR TITLE
Task annotations

### PR DESCRIPTION
**Pull request by geohacker**
Work in progress to address https://github.com/hotosm/tasking-manager/issues/1284

POST /project/<id>/task_annotations/<type>
example /project/1/task_annotations/ml
post body:
```
{
    'projectId': 1,
    'tasks': [
        {
            'taskId': 1,
            'annotationSource': 'https://some-url.com/ml',
            'annotationType': 'ml',
            'properties': {
                'building_area_ml_pred': 0.7,
                'building_area_osm': 0.8
            }
        },
        {
            'taskId': 2,
            'annotationSource': 'https://some-url.com/ml',
            'annotationType': 'ml',
            'properties': {
                'building_area_ml_pred': 0.4,
                'building_area_osm': 0.2
            }
        }
    ]
 }
```

GET /project/<id>/task_annotations/<type>
example /project/1/task_annotations/ml
response:

```
{
    "projectId": 1,
    "annotationType": "ml",
    "tasks": [
        {
            "taskId": 1,
            "annotationType": "ml",
            "annotationSource": null,
            "properties": {
                "building_area_ml_pred": 0,
                "building_area_osm": 0.8
            }
        },
        {
            "taskId": 2,
            "annotationType": "ml",
            "annotationSource": null,
            "properties": {
                "building_area_ml_pred": 0,
                "building_area_osm": 0.2
            }
        }
    ]
}
```

cc @ethan-nelson @wronk @batpad @smit1678

### CODE REVIEW

**Comment by ethan-nelson** [comment]
Thanks @geohacker I see that now. Feel free to ping me when you are ready for a review

File: client/app/services/style.service.js [hunk 1]
```diff
@@ -43,14 +43,25 @@
         // The get style functions could check the cache for the style and if found return it, otherwise
         // create a new one and add to cache
         // var styleCache = {};
+        var d3Scale = null;
+
+        function setD3Scale(domain) {
+            if (!d3Scale) {
+                d3Scale = d3.scaleQuantile().domain(domain).range(d3.schemeReds[5]);
```
Comment (wronk) [inline client/app/services/style.service.js]: We may want to bring this down to 4 or 3 eventually, but this is okay for now

File: server/api/project_apis.py [hunk 2]
```diff
@@ -440,3 +443,128 @@ def get(self, project_id: int):
...
+              required: true
+              type: string
+              default: application/json
+            - name: project_id
+              in: path
+              description: The unique project ID
+              required: true
+              type: integer
+              default: 1
+            - in: body
```
Comment (wronk) [inline server/api/project_apis.py]: I'm no good at JS, but is there a mismatch here between `body` and `annotation_str`?

File: client/app/services/style.service.js (hunk 1 above)
Comment (geohacker) [reply client/app/services/style.service.js]: 👍

File: server/api/project_apis.py (hunk 2 above)
Comment (geohacker) [reply server/api/project_apis.py]: @wronk I used another endpoint as an example to write the documentation string - I think that just says it's in the body of the request - should be ok.

**Comment by geohacker** [comment]
@wronk  @ethan-nelson- I think this is ready for a review. Outstanding item is to add interaction when we visualize the ML annotations, and setting up the legend. We can get to it if there are no flags on this approach. Note that this is a better version of what we built in our fork.

File: server/api/project_apis.py (hunk 2 above)
Comment (ethan-nelson) [reply server/api/project_apis.py]: I think `body` is okay for the JSON data, but we should have another item in this comment block for `annotation_type` so swagger-ui can parse it as a field to include on the api page.

File: server/api/project_apis.py [hunk 3]
```diff
@@ -440,3 +443,128 @@ def get(self, project_id: int):
...
+
+        # check if task ids are valid
+        tasks = Task.get_tasks(project_id, task_ids)
+        tasks_ids_db = [t.id for t in tasks]
+        if (len(task_ids) != len(tasks_ids_db)):
+            return {"error": 'Invalid task id'}, 500
+
+        for annotation in annotations['tasks']:
+            TaskAnnotationsService.add_or_update_annotation(annotation, project_id, annotation_type)
+
```
Comment (ethan-nelson) [inline server/api/project_apis.py]: Can we add an HTTP 200 return if the add/update goes successfully?

**Comment by ethan-nelson** [comment]
Thanks for your work bringing this together! I gave it an initial look and added a few individual comments to the code. However, I cannot get the annotations to actually display.

I have a simple test project with four tasks and I have (successfully) posted:

```
{"tasks": [
  {"taskId": 1, "properties": {"building_area_diff": "#FFF"}},
  {"taskId": 2, "properties": {"building_area_diff": "#311"}},
  {"taskId": 3, "properties": {"building_area_diff": "#4B7"}},
  {"taskId": 4, "properties": {"building_area_diff": "#299"}}
]}
```

I verify that this is stored successfully in the database. Once I load the project page on my machine and click the "ML Layer" button, I receive an angular error:

```
TypeError: $.projectData.annotations is undefined
Stack trace:
e/$.addAnnotationsToMap@http://192.168.1.6:4000/app/taskingmanager.min.js:1:30858
fn@https://cdnjs.cloudflare.com/ajax/libs/angular.js/1.6.3/angular.min.js line 241 > Function:4:368
e@https://cdnjs.cloudflare.com/ajax/libs/angular.js/1.6.3/angular.min.js:285:410
$eval@https://cdnjs.cloudflare.com/ajax/libs/angular.js/1.6.3/angular.min.js:149:176
$apply@https://cdnjs.cloudflare.com/ajax/libs/angular.js/1.6.3/angular.min.js:149:400
compile/</<@https://cdnjs.cloudflare.com/ajax/libs/angular.js/1.6.3/angular.min.js:285:460
hg@https://cdnjs.cloudflare.com/ajax/libs/angular.js/1.6.3/angular.min.js:39:297
d@https://cdnjs.cloudflare.com/ajax/libs/angular.js/1.6.3/angular.min.js:39:246
```

Here is the ml task-annotations that was loaded by angular:

```
{
    "projectId": 3619,
    "tasks": [
        {
            "taskId": 1,
            "annotationType": "ml",
            "annotationSource": null,
            "properties": {
                "building_area_diff": "#FFF"
            }
        },
        {
            "taskId": 2,
            "annotationType": "ml",
            "annotationSource": null,
            "properties": {
                "building_area_diff": "#311"
            }
        },
        {
            "taskId": 3,
            "annotationType": "ml",
            "annotationSource": null,
            "properties": {
                "building_area_diff": "#4B7"
            }
        },
        {
            "taskId": 4,
            "annotationType": "ml",
            "annotationSource": null,
            "properties": {
                "building_area_diff": "#299"
            }
        }
    ]
}
```

Perhaps I have an incorrect assumption about the data expected. If so, could you provide a sample with N sequentially numbered tasks for a project that I can post to one locally for testing?

File: server/api/project_apis.py [hunk 4]
```diff
@@ -440,3 +443,128 @@ def get(self, project_id: int):
             error_msg = f'Project Summary GET - unhandled error: {str(e)}'
             current_app.logger.critical(error_msg)
             return {"error": error_msg}, 500
+
+
+class TaskAnnotationsAPI(Resource):
+    def get(self, project_id: int, annotation_type: str = None):
```
Comment (ethan-nelson) [inline server/api/project_apis.py]: I couldn't get the GET or POST calls to work successfully without a parameter. If that is a blocking point for now, we can drop it and work on it later when broadening the capabilities of the task annotations (such as the property fields to load in the project).

File: client/app/project/project.controller.js [hunk 5]
```diff
@@ -403,9 +404,9 @@
                 addAoiToMap(vm.projectData.areaOfInterest);
                 addPriorityAreasToMap(vm.projectData.priorityAreas);
                 addProjectTasksToMap(vm.projectData.tasks, true);
+                loadAnnotations(id);
```
Comment (ethan-nelson) [inline client/app/project/project.controller.js]: I know you're still working on the front end side, but figured I would comment this while it's fresh.

Loading on a project with stored annotations works fine and is hardly noticeable in that process. However, if a project does _not_ have annotations, the 404 thrown doesn't seem to be caught well by the frontend and it thinks there is no project data, causing e.g. the project text to revert to `undefined`--until the next refresh of project data (without refresh of  annotations) after 10 seconds.

File: server/api/project_apis.py (hunk 2 above)
Comment (geohacker) [reply server/api/project_apis.py]: Done!

**Comment by geohacker** [comment]
@ethan-nelson thanks for the review!

> However, I cannot get the annotations to actually display.

The payload is incorrect, sorry I should have pasted an example. The annotation data isn't expected to be colour codes. Instead it should just be floats like [the example here](https://github.com/hotosm/tasking-manager/pull/1289#issue-236500669). We put them on a colour scale before visualising on the front-end. Here's an example you can use to test https://gist.github.com/geohacker/60ce62fed060ff8b398b4961210f633d

File: client/app/project/project.controller.js (hunk 5 above)
Comment (geohacker) [reply client/app/project/project.controller.js]: Done

**Comment by dakotabenjamin** [comment]
I changed the base branch to `tm4ml`. There aren't any conflicts to feel free to merge at will.

**Comment by geohacker** [comment]
@dakotabenjamin thank you!

@ethan-nelson - I pulled the auth-for-applications branch into this one and integrated the check for `Application-Token` in the header. Let me know what you think. I think we can merge this and stage. 

cc @wronk @smit1678

**Pull request merged**